lib/client/api.go

```diff
@@ -699,6 +700,37 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error,
 		return trace.Wrap(err)
 	}
 
+	// The user has typed a command like `tsh ssh ...` without being logged in,
+	// if the running binary needs to be updated, update and re-exec.
+	//
+	// If needed, download the new version of {tsh, tctl} and re-exec. Make
+	// sure to exit this process with the same exit code as the child process.
+	//
+	toolsDir, err := tools.Dir()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	updater := tools.NewUpdater(tools.DefaultClientTools(), toolsDir, teleport.Version)
+	toolsVersion, reExec, err := updater.CheckRemote(ctx, tc.WebProxyAddr, tc.InsecureSkipVerify)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if reExec {
+		// Download the version of client tools required by the cluster.
+		err := updater.UpdateWithLock(ctx, toolsVersion)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			utils.FatalError(err)
+		}
+		// Re-execute client tools with the correct version of client tools.
+		code, err := updater.Exec()
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
```

Comment:
Do we need a warning or debug log if ErrNotExist?

Reply:
I think better not add it, this check verify if the binary was created under `~/.tsh/bin/{tsh,tctl}`, this one specific case when we started update, but then user send interrupt signal to cancel update (with debug log we going to message it every time), so we need to continue execution in current version